### PR TITLE
Implement mono_gc_alloc_fixed/mono_gc_free_fixed on Boehm to match SGen

### PR DESCRIPTION
**Pull request by joncham**
Implement mono_gc_alloc_fixed on Boehm to be uncollectable. This matches SGen behavior and avoids needing to register fixed allocated memory as a root.

### CODE REVIEW

File: mono/metadata/domain.c [hunk 1]
```diff
@@ -382,18 +382,7 @@ mono_domain_create (void)
...
-	/*
-	 * Boehm doesn't like roots inside GC allocated objects, and alloc_fixed returns
-	 * a GC_MALLOC-ed object, contrary to the api docs. This causes random crashes when
-	 * running the corlib test suite.
-	 * To solve this, we pass a NULL descriptor, and don't register roots.
-	 */
-	domain = (MonoDomain *)mono_gc_alloc_fixed (sizeof (MonoDomain), NULL, MONO_ROOT_SOURCE_DOMAIN, "domain object");
-#else
 	domain = (MonoDomain *)mono_gc_alloc_fixed (sizeof (MonoDomain), domain_gc_desc, MONO_ROOT_SOURCE_DOMAIN, "domain object");
-	mono_gc_register_root ((char*)&(domain->MONO_DOMAIN_FIRST_GC_TRACKED), G_STRUCT_OFFSET (MonoDomain, MONO_DOMAIN_LAST_GC_TRACKED) - G_STRUCT_OFFSET (MonoDomain, MONO_DOMAIN_FIRST_GC_TRACKED), MONO_GC_DESCRIPTOR_NULL, MONO_ROOT_SOURCE_DOMAIN, "misc domain fields");
```
Comment (vargaz) [inline mono/metadata/domain.c]: Why is this removed ?

Comment (joncham) [inline mono/metadata/domain.c]: I didn't see why it was needed if the `domain` was already allocated with fixed memory using a descriptor to mark those fields?

Comment (joncham) [inline mono/metadata/domain.c]: I thought the descriptor we build a few lines above served the same purpose, of indicating that memory should be scanned by the GC.

Comment (vargaz) [inline mono/metadata/domain.c]: The descriptor only applies to fields between FIRST_OBJECT and FIRST_GC_TRACKED.

Comment (joncham) [inline mono/metadata/domain.c]: Ah, I see the difference. I will revert this part of the patch.

Comment (joncham) [inline mono/metadata/domain.c]: Actually, it looks like everything but one value between `MONO_DOMAIN_FIRST_GC_TRACKED` and `MONO_DOMAIN_FIRST_GC_TRACKED` is a `MonoGHashTable`. The remaining item is the root `static_data_array` for domain statics which is allocated with `mono_gc_alloc_fixed`.

I can remove this change for now (and open another PR that cleans it up further), but don't think any of these items need scanned by GC as all should be immovable memory already tracked by the GC.

**Comment by joncham** [comment]
I removed the changes affect the non-object reference fields of the MonoDomain being registered as a root.

**Comment by joncham** [comment]
Actually, new a few more fixes. Please don't merge yet...

**Comment by vargaz** [comment]
Looks ok.
